drivers: sensor: basic tcs3472x support

thokis opened the pull request:
Inspired largely by the [Adafruit TCS34725 Color Sensor Driver](https://github.com/adafruit/Adafruit_TCS34725#adafruit-tcs34725-color-sensor-driver)

- [x] basic AMS TCS3472x support
- [ ] calculate `lux` value for  [`SENSOR_CHAN_LIGHT`](https://docs.zephyrproject.org/latest/hardware/peripherals/sensor.html#c.sensor_channel.SENSOR_CHAN_LIGHT)
- [ ] power management
- [ ] interrupt handling via [`Triggers`](https://docs.zephyrproject.org/latest/hardware/peripherals/sensor.html#triggers)

alexanderwachter commented:
@thokis thanks for your PR.
There might be some style errors. I triggered the CI to highlite them. Can you split the PR into at least two commits?
Driver and sample, for example.

Commit:
drivers: sensor: added tcs3472x support

Initial support for the AMS TCS3472X color sensor.

Signed-off-by: Thomas Kiss <thomas.kiss@lemonbeat.com>

Commit:
samples: sensor: tcs3472x sample added

Added a sample for the tcs3472x driver. Tested with
`nucleo_f429zi`, `nrf9160dk_nrf9160_ns` and `esp32` boards.

Signed-off-by: Thomas Kiss <thomas.kiss@lemonbeat.com>

thokis commented:
@alexanderwachter i hope this is better now.

thokis commented:
I think it is better to mark this MR as Draft, because the driver is not really feature complete.

alexanderwachter commented:
You can. But you can also try to get this basic version merged and make a follow-up PR withe PM and interrupt handling.

Your commit messages need a signed off (use git commit -s) and a body.
The body needs a separation wit an empty line.
Please refer to the contribution guide for the CI errors.
There are some helpful tools you can run before pushing, like gitlint and checkpatch.

thokis commented:
@alexanderwachter That sounds logical. Should i then remove the checkboxes in the MR description?

alexanderwachter commented:
Nah, does not hurt.
Usually we use issues to track those, but in this case it would be a bit too much.

thokis commented:
@alexanderwachter Sorry for bothering you but could you explain me [this](https://github.com/zephyrproject-rtos/zephyr/actions/runs/5584943608/jobs/10208965136?pr=60466) failed test?

Especially the line 49 `/runner/_work/zephyr/zephyr/doc/_build/src/samples/sensor/tcs3472x/README.rst:9: ERROR: Unknown target name: "ams tcs3472x".` is a little bit confusing for me. I presume i have to add some entry into somewhere in `/docs` but  i'm not really sure.

MaureenHelm commented on drivers/sensor/tcs3472x/Kconfig:
```diff
@@ -0,0 +1,67 @@
+	bool "TCS3472X sensor"
+	default y
+	depends on DT_HAS_AMS_TCS3472X_ENABLED
+	select I2C if $(dt_compat_on_bus,$(DT_COMPAT_AMS_TCS3457X),i2c)
```
You can simplify this line since the sensor only supports I2C (no SPI or other interfaces).
```suggestion
	select I2C
```

MaureenHelm commented on drivers/sensor/tcs3472x/tcs3472x.c:
```diff
@@ -0,0 +1,216 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/init.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/__assert.h>
+
+#include <zephyr/logging/log.h>
+
+#include "tcs3472x.h"
+#include "zephyr/drivers/i2c.h"
+#include <stdint.h>
+#include <stdlib.h>
```
I think you can remove gpio, stdint, and stdlib. Please also change i2c to a system include.

MaureenHelm commented on drivers/sensor/tcs3472x/tcs3472x.c:
```diff
@@ -0,0 +1,216 @@
+	uint8_t chip_id;
+};
+
+int tcs3472x_reg_read(const struct i2c_dt_spec *i2c, uint8_t reg, uint8_t *buf, uint8_t size)
```
Make all these functions static

MaureenHelm commented on drivers/sensor/tcs3472x/tcs3472x.c:
```diff
@@ -0,0 +1,216 @@
+	if (ret < 0) {
+		return -EIO;
+	}
+	k_sleep(K_USEC(2400));
```
Please add a comment documenting why you chose this sleep duration.

MaureenHelm commented on drivers/sensor/tcs3472x/tcs3472x.c:
```diff
@@ -0,0 +1,216 @@
+	if (ret < 0) {
+		return -EIO;
+	}
+	k_sleep(K_MSEC(CONFIG_TCS3472X_INT_TIME_SCALE * 12 / 5 + 1));
```
Please add a comment documenting why you chose this sleep duration.

MaureenHelm commented on drivers/sensor/tcs3472x/tcs3472x.c:
```diff
@@ -0,0 +1,216 @@
+	const struct tcs3472x_config *config = dev->config;
+	uint8_t id = 0U;
+
+	if (!device_is_ready(config->i2c.bus)) {
```
Use `i2c_is_ready_dt`

MaureenHelm commented on drivers/sensor/tcs3472x/tcs3472x.c:
```diff
@@ -0,0 +1,216 @@
+			      &tcs3472x_data_##inst,			\
+			      &tcs3472x_config##inst,			\
+			      POST_KERNEL,				\
+			      CONFIG_APPLICATION_INIT_PRIORITY,		\
```
Use `CONFIG_SENSOR_INIT_PRIORITY`

MaureenHelm commented on samples/sensor/tcs3472x/README.rst:
```diff
@@ -0,0 +1,99 @@
+.. _tcs3472x:
+
+TCS3472x Color Sensor
```
Please look at making this driver generic for any color sensor. See `samples/sensor/accel_polling` for example

thokis replied:
Hi, do your mean that i should rewrite the sample or the driver itself?

thokis replied:
I like your idea of rewriting the sample, maybe i could test it with an additional APDS9960 from Adafruit or so. Wouldn't it be better that this MR should only contain the driver component then?

MaureenHelm replied:
Sorry for the confusing typo in my earlier comment. I meant to say that the sample application should be generic, not the driver itself. It used to be the case that we had multiple accelerometer samples, multiple magnetometer samples, multiple temperature sensor samples, etc. and it got out of control having a new sample added for every new sensor driver, so we have been trying to simplify down to one sample per type of sensor.

If you happen to have another color sensor on hand to test the generic sample that would be great, but not required.

MaureenHelm reviewed (changes requested):
Thanks for your contribution!

Please add this sensor to `tests/drivers/build_all/sensor/i2c.dtsi` to ensure it builds in CI.

alexanderwachter commented on drivers/sensor/tcs3472x/tcs3472x.c:
```diff
@@ -0,0 +1,216 @@
+		val->val2 = data->c_chan;
+		break;
+	case SENSOR_CHAN_RED:
+		val->val1 = (uint16_t)((float)(data->r_chan) / data->c_chan * 255) / 100;
```
I would not use float.
If a user calls this function without float support enabled in his threads, a race condition could happen.

thokis replied:
Removed `float` usage

Commit:
drivers: sensor: updated tcs3472x driver

This commit moves the configuration of the ADC integration time and ADC
gain from Kconfig to the device-tree.

Signed-off-by: Thomas Kiss <thomas.kiss@lemonbeat.com>

thokis committed:
Merge branch 'zephyrproject-rtos:main' into tcs3472x
